### PR TITLE
txn: use region pessimsitic lock rollback to speed up cleanup

### PR DESCRIPTION
**Pull request by cfzjywxk**
Ref: https://github.com/tikv/tikv/issues/16158

Use region pessimistic rollback to speed up expired pessimistic lock cleanup, it would take effect for:
1. write-write conflict pessimistic lock resolve, for example when a pessimistic lock is encountered by a prewrite request or another pessimistic request.

It would not impact:
1. GC lock resolving as it should ensure no expired locks are left.
2. Other lock resoving like `snapshot.scan`.

---

To be discussed, if it's necesary to removing `for_update_ts` check.

The benefit of removing  `for_update_ts` check in tikv:
All the expired locks belonging to the same transction could be cleaned up at one time.

The side effect of removing  `for_update_ts` check in tikv:
When the primary lock key is changed (e.g the first statement fails and the related pessimitic locks are left), it's possble the normal pessimistic locks are cleaned up mistakenly if https://github.com/tikv/tikv/pull/16185#discussion_r1436859199 `for_update_ts` check is removed.

### CODE REVIEW

File: txnkv/txnlock/lock_resolver.go [hunk 1]
```diff
@@ -350,11 +356,12 @@ func (lr *LockResolver) BatchResolveLocks(bo *retry.Backoffer, locks []*Lock, lo
...
-	Locks         []*Lock
-	Lite          bool
-	ForRead       bool
-	Detail        *util.ResolveLockDetail
+	CallerStartTS   uint64
+	Locks           []*Lock
+	Lite            bool
+	ForRead         bool
+	Detail          *util.ResolveLockDetail
+	PessimisticLite bool
```
Comment (zyguan) [inline txnkv/txnlock/lock_resolver.go:364]: I'm thinking about: do we realy need `PessimisticLite`? can we just use `Lite` for both optimistic and pessimistic.

**Review by zyguan — APPROVED** [review]

File: txnkv/txnlock/lock_resolver.go (hunk 1 above)
Comment (cfzjywxk) [reply txnkv/txnlock/lock_resolver.go]: Yes, I'am thinking about risks about using region resolve for all the expired pessimistic locks, if there is some situations cluster stability would be afffected.

File: txnkv/txnlock/lock_resolver.go [hunk 2]
```diff
@@ -1178,7 +1197,10 @@ func (lr *LockResolver) resolveLock(bo *retry.Backoffer, l *Lock, status TxnStat
 
 // resolvePessimisticLock handles pessimistic locks after checking txn status.
 // Note that this function assumes `CheckTxnStatus` is done (or `getTxnStatusFromLock` has been called) on the lock.
-func (lr *LockResolver) resolvePessimisticLock(bo *retry.Backoffer, l *Lock) error {
+// When "lite" is set to true, only pessimistic rollback input lock. When "lite" is set to false, the corresponding
```
Comment (zyguan) [inline txnkv/txnlock/lock_resolver.go:1200]: lite -> pessimisticRegionResolve

**Review by zyguan — APPROVED** [review]

File: txnkv/txnlock/lock_resolver.go [hunk 3]
```diff
@@ -261,7 +271,7 @@ func (lr *LockResolver) BatchResolveLocks(bo *retry.Backoffer, locks []*Lock, lo
 			//
 			// `resolvePessimisticLock` should be called after calling `getTxnStatus`.
 			// See: https://github.com/pingcap/tidb/issues/45134
-			err := lr.resolvePessimisticLock(bo, l)
+			err := lr.resolvePessimisticLock(bo, l, false, pessimisticCleanRegions)
```
[the remark this answers is not in the document]
Comment (zyguan) [reply txnkv/txnlock/lock_resolver.go:274]: If region 1 and region 2 get merged, then the RegionVerID changed?

File: txnkv/txnlock/lock_resolver.go [hunk 4]
```diff
@@ -160,6 +160,10 @@ func (s TxnStatus) StatusCacheable() bool {
 	return false
 }
 
+func (s TxnStatus) String() string {
+	return fmt.Sprintf("ttl:%v commit_ts:%v action: %v, primary_lock:%v", s.ttl, s.commitTS, s.action, s.primaryLock)
```
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply txnkv/txnlock/lock_resolver.go]: The `primaryLock` is removed from printing.

File: txnkv/txnlock/lock_resolver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply txnkv/txnlock/lock_resolver.go]: The usage for `BatchResolve` is removed, I've added some more details in the description, PTAL.

**Comment by cfzjywxk** [comment]
Why the integration test cases could not pass on the CI..

**Comment by cfzjywxk** [comment]
@zyguan @MyonKeminta 
At present, the CI integration test fails when the region resolve is enabled for the prewrite and pessimistic commands during lock resolving. Strangely, the test passes on my local machine. While the feature is not currently enabled, we might proceed with the PR review. 

I plan to investigate the integration test failure on CI and, if needed, implement the region resolve on the mockstore to ensure its success.

**Review by MyonKeminta — APPROVED** [review]

**Review by zyguan — APPROVED** [review]

**Pull request merged**